modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/bookmark/components/UpsellViewModel.kt

```diff
@@ -0,0 +1,87 @@
+package au.com.shiftyjelly.pocketcasts.player.view.bookmark.components
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import au.com.shiftyjelly.pocketcasts.analytics.AnalyticsEvent
+import au.com.shiftyjelly.pocketcasts.analytics.AnalyticsTrackerWrapper
+import au.com.shiftyjelly.pocketcasts.analytics.SourceView
+import au.com.shiftyjelly.pocketcasts.models.type.Subscription
+import au.com.shiftyjelly.pocketcasts.models.type.Subscription.SubscriptionTier
+import au.com.shiftyjelly.pocketcasts.models.type.SubscriptionMapper
+import au.com.shiftyjelly.pocketcasts.repositories.subscription.ProductDetailsState
+import au.com.shiftyjelly.pocketcasts.repositories.subscription.SubscriptionManager
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.reactive.asFlow
+import javax.inject.Inject
+
+@HiltViewModel
+class UpsellViewModel @Inject constructor(
+    private val analyticsTracker: AnalyticsTrackerWrapper,
+    private val subscriptionManager: SubscriptionManager,
+) : ViewModel() {
+
+    private val _state: MutableStateFlow<UiState> = MutableStateFlow(UiState.Loading)
+    val state: StateFlow<UiState> = _state
+
+    init {
+        viewModelScope.launch {
+            subscriptionManager
+                .observeProductDetails()
+                .asFlow()
+                .stateIn(viewModelScope)
+                .collect { productDetails ->
+                    val subscriptions = when (productDetails) {
+                        is ProductDetailsState.Error -> null
+                        is ProductDetailsState.Loaded -> productDetails.productDetails.mapNotNull { productDetailsState ->
+                            Subscription.fromProductDetails(
+                                productDetails = productDetailsState,
+                                isFreeTrialEligible = subscriptionManager.isFreeTrialEligible(
+                                    SubscriptionMapper.mapProductIdToTier(productDetailsState.productId)
+                                )
+                            )
+                        }
+                    } ?: emptyList()
+                    updateState(subscriptions)
+                }
+        }
+    }
+
+    private fun updateState(
+        subscriptions: List<Subscription>,
+    ) {
+        val subscriptionTier = SubscriptionTier.PATRON
+        val updatedSubscriptions = subscriptions.filter { it.tier == subscriptionTier }
+
+        val selectedSubscription = subscriptionManager.getDefaultSubscription(
```

Comment:
It would be good to add a comment here as it took me a while to figure out this logic figured out if the tier had a free trial. Something like:
```
// check the server subscriptions to see if the Patron tier has a free trial
```

Reply:
Good point! Added in b49d402d6324ede717fabdea95d97fe6adfe5394.

Reply:
Thanks